V4 Updates for Metrics, Tracing and Security (#336)
SteeltoeOSS/Documentation: api/v4/tracing/index.md

# Distributed Tracing

On the subject of distributed tracing for .NET applications, previous versions of Steeltoe offered either an implementation of OpenCensus or shortcuts for enabling distributed tracing with [OpenTelemetry](https://opentelemetry.io/).
As of this writing, the Steeltoe team believe that OpenTelemetry has evolved to the point that a major Steeltoe component is no longer necessary, and we recommend using OpenTelemetry directly.

Steeltoe continues to directly offer an option for [log correlation](#log-correlation). This page also provides direction for developers looking to achieve the same outcomes Steeltoe has previously provided more directly.

## Introduction

As the name implies, distributed tracing is a way of tracing requests through distributed systems.
Distributed tracing is typically accomplished by instrumenting components of the system to recognize and pass along metadata that is specific to a particular action or user request, and using another backend system to reconstruct the flow of the request through that metadata.

> In the parlance of distributed tracing, a "span" is the basic unit of work. For example, sending an RPC is a new span, as is sending a response to an RPC.

Reviewer: ```suggestion
In the parlance of distributed tracing, a "span" is the basic unit of work. For example, sending an HTTP request creates a new span, as does sending a response.
```

> Spans are identified by a unique ID for the span and by another ID for the "trace", which the span is part of.

Reviewer: ```suggestion
Spans are identified by a unique ID and contain the ID for the "trace" it is part of.
```

> Spans also have other data like descriptions, key-value annotations, the ID of the span that caused them, and process IDs.

Reviewer: ```suggestion
Spans also have other data like descriptions, key-value annotations, the ID of the span that initiated the execution flow, and process IDs.
```

Spans are started and stopped, and they keep track of their timing information. Once you create a span, you must stop it at some point in the future.

> A set of spans form a tree-like structure called a "trace". For example, a trace might be formed by a POST request that adds an item to a shopping cart.

Reviewer: ```suggestion
A set of spans form a tree-like structure called a "trace". For example, a trace might be formed by a POST request that adds an item to a shopping cart, which results in calling several backend services.
```

## Log correlation

Log correlation is all about taking log entries from disparate systems and bringing them together using some matching criteria (such as a distributed trace ID).
The process can be easier when important pieces of data are logged in the same format across different systems (such as .NET and Java apps communicating with each other).

Steeltoe provides the class `TracingLogProcessor`, which is an `IDynamicMessageProcessor` for correlating logs. The processor is built for use with a [Steeltoe Dynamic Logging provider](../logging/index.md).

> The result is log entries that use the same trace format popularized by [Spring Cloud Sleuth](https://cloud.spring.io/spring-cloud-sleuth/reference/html/#log-correlation),

Reviewer: ```suggestion
It enriches log entries with correlation data using the same trace format popularized by [Spring Cloud Sleuth](https://cloud.spring.io/spring-cloud-sleuth/reference/html/#log-correlation),
```

The only remaining step is to register the processor:

```csharp
using Steeltoe.Management.Tracing;

var builder = WebApplication.CreateBuilder(args);
builder.Services.AddTracingLogProcessor();
```

> [!NOTE]
> This extension method also ensures that implementations of `IApplicationInstanceInfo` and `IDynamicLoggerProvider` have been registered.
> If you wish to customize either of those or use non-default implementations, do so before calling `AddTracingLogProcessor`.

## OpenTelemetry

To use OpenTelemetry, you _can_ start by adding a reference to the `OpenTelemetry.Extensions.Hosting` NuGet package.
This package provides access to `OpenTelemetryBuilder`, which is the main entrypoint to OpenTelemetry, but depending on which packages you add later you can probably skip this step.
The rest of this section contains information on pieces of OpenTelemetry that could be configured by previous versions of Steeltoe.

### Sampler configuration

OpenTelemetry Provides the `Sampler` abstraction for configuring when traces should be recorded.
The simplest options are `AlwaysOnSampler` and `AlwaysOffSampler`, with their names describing exactly which traces will be recorded.

> To replace the Steeltoe configuration for using these samplers, set the environment variable `OTEL_TRACES_SAMPLER` to `always_on` or `always_off`.

Reviewer: ```suggestion
As a replacement for what Steeltoe used to provide for using these samplers, set the environment variable `OTEL_TRACES_SAMPLER` to `always_on` or `always_off`.
```

> [!TIP]
> OpenTelemetry is generally built to follow the [options pattern](https://learn.microsoft.com/dotnet/core/extensions/options).

> > There are more ways to configure options than what is demonstrated on this page, these are only examples of the potential.

Reviewer: ```suggestion
> There are more ways to configure options than demonstrated on this page, these are just examples to get started.
```

### ASP.NET Core

To instrument requests coming into the application through ASP.NET Core, start by adding a reference to the `OpenTelemetry.Instrumentation.AspNetCore` NuGet package.

Next, add the instrumentation to the `TracerProviderBuilder`:

```csharp
services.AddOpenTelemetry().WithTracing(tracerProviderBuilder => tracerProviderBuilder.AddAspNetCoreInstrumentation());
```

In order to replicate the Steeltoe setting `IngressIgnorePattern` (a Regex pattern describing which incoming requests to ignore), configure the `AspNetCoreTraceInstrumentationOptions`:

```csharp
private const string DefaultIngressIgnorePattern = "/actuator/.*|/cloudfoundryapplication/.*|.*\\.png|.*\\.css|.*\\.js|.*\\.html|/favicon.ico|/hystrix.stream|.*\\.gif";
private static readonly Regex IngressPathMatcher = new(DefaultIngressIgnorePattern, RegexOptions.Compiled | RegexOptions.CultureInvariant, TimeSpan.FromSeconds(1));

services.PostConfigure<AspNetCoreTraceInstrumentationOptions>(aspNetCoreTraceInstrumentationOptions =>
{
    aspNetCoreTraceInstrumentationOptions.Filter += httpContext => !IngressPathMatcher.IsMatch(httpContext.Request.Path);
});

> ```

Reviewer: The Regex should just be a variable, and the call should take place on `builder.Services`. I think `aspNetCoreTraceInstrumentationOptions` should just be renamed to `options` here.

# Distributed Tracing

> On the subject of distributed tracing for .NET applications, previous versions of Steeltoe offered either an implementation of OpenCensus or shortcuts for enabling distributed tracing with [OpenTelemetry](https://opentelemetry.io/).
> By now, OpenTelemetry has evolved to the point that a Steeltoe component is no longer necessary, and we recommend using OpenTelemetry directly.

Reviewer: ```suggestion
On the subject of distributed tracing for .NET applications, previous versions of Steeltoe offered either an implementation of OpenCensus or shortcuts for enabling distributed tracing with [OpenTelemetry](https://opentelemetry.io/) to integrate with the Tanzu platform. We painted ourselves into a corner with Steeltoe 3, which has a strong dependency on OpenTelemetry packages. New versions of OpenTelemetry contained breaking changes, which we couldn't adapt to without resorting to reflection. While we do that in other places, reflection is slow and tracing is performance-sensitive.

By now, OpenTelemetry has evolved to the point that only a few lines of code are needed. So instead of providing a Steeltoe component, the guidance below is offered to accomplish the same. The benefit of that is greater flexibility: when OpenTelemetry changes, there's no need to wait for a new Steeltoe version first.
```

As a replacement for what Steeltoe used to provide for using these samplers, set the environment variable `OTEL_TRACES_SAMPLER` to `always_on` or `always_off`.

> [!TIP]
> OpenTelemetry is generally built to follow the [options pattern](https://learn.microsoft.com/dotnet/core/extensions/options).
> There are more ways to configure options than demonstrated on this page, these are just examples to get started.

### Set Application Name

In order to use the Steeltoe name for your application with OpenTelemetry, call `SetResourceBuilder` and pass in a value from the registered `IApplicationInstanceInfo`:

```csharp
builder.Services.ConfigureOpenTelemetryTracerProvider((serviceProvider, tracerProviderBuilder) =>
{
    var appInfo = serviceProvider.GetRequiredService<IApplicationInstanceInfo>();
    tracerProviderBuilder.SetResourceBuilder(ResourceBuilder.CreateDefault().AddService(appInfo.ApplicationName!));
});
```

The above example assumes you are already using some other Steeltoe component which adds `IApplicationInstanceInfo` to the IoC container. If that is not the case, there are two steps to take to register the default implementation:

> 1. Add a NuGet package reference to `Steeltoe.Common`

Author: If the page is followed completely as a step-by-step, yes... But Steeltoe.Management.Tracing isn't required for anything under `## OpenTelemetry`

1. Call `AddApplicationInstanceInfo`

```csharp
using Steeltoe.Common.Extensions;

var builder = WebApplication.CreateBuilder(args);

builder.Services.AddApplicationInstanceInfo();
```

> ## Instrumenting applications

Reviewer: It's unclear to me how the sections on this page relate. For example, "OpenTelemetry" is at H2, and so is this one. That implies it is independent of the previous section. But on the other hand, the code fragments here build on top of the previous section, which is the case for most individual H2 chapters.

Sections at the same depth should not depend on previous sections, so that users can jump in without starting from the top of the page.

Author: Good catch, there should only be 2 H2 on this page

    Regex ingressPathMatcher = new(defaultIngressIgnorePattern, RegexOptions.Compiled | RegexOptions.CultureInvariant, TimeSpan.FromSeconds(1));
    options.Filter += httpContext => !ingressPathMatcher.IsMatch(httpContext.Request.Path);
});
```

As an alternative to using a regular expression, you can list out the paths to ignore in the Filter property (`Filter` is a `Func<HttpContext, bool>?`):

```csharp
services.PostConfigure<AspNetCoreTraceInstrumentationOptions>(aspNetCoreTraceInstrumentationOptions =>
{
    aspNetCoreTraceInstrumentationOptions.Filter += httpContext =>
        !httpContext.Request.Path.StartsWithSegments("/actuator", StringComparison.OrdinalIgnoreCase) &&
        !httpContext.Request.Path.StartsWithSegments("/cloudfoundryapplication", StringComparison.OrdinalIgnoreCase);
});
```

> [!TIP]
> By default, the ASP.NET Core instrumentation does not filter out any requests.
> The latter approach above may quickly prove unwieldy if there are many patterns to ignore, such as when listing many file types (notably missing from the example).

> [Learn more about ASP.NET Core instrumentation for OpenTelemetry](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Instrumentation.AspNetCore)

Reviewer: Alternatively, include the link in the introduction.

## Propagating Trace Context

By default, OpenTelemetry uses the [W3C trace context](https://github.com/w3c/trace-context) for propagating traces.
Some systems like Cloud Foundry may still be configured for the Zipkin standard of [B3 propagation](https://github.com/openzipkin/b3-propagation).

In order to use B3 propagation, add a reference to the `OpenTelemetry.Extensions.Propagators` NuGet package.

Next, let the compiler know that the `B3Propagator` should come from the package reference you just added (rather than the deprecated class found in `OpenTelemetry.Context.Propagation`):

```csharp
using B3Propagator = OpenTelemetry.Extensions.Propagators.B3Propagator;
```

Finally, register a `CompositeTextMapPropagator` that includes the `B3Propagator` and `BaggagePropagator`:

```csharp
builder.Services.ConfigureOpenTelemetryTracerProvider((serviceProvider, tracerProviderBuilder) =>
{
    List<TextMapPropagator> propagators =
    [
        new B3Propagator(),
        new BaggagePropagator()
    ];

    Sdk.SetDefaultTextMapPropagator(new CompositeTextMapPropagator(propagators));
});
```

> By default, `B3Propagator` uses [multiple headers](https://github.com/openzipkin/b3-propagation?tab=readme-ov-file#multiple-headers). In order to use a [single B3 header](https://github.com/openzipkin/b3-propagation?tab=readme-ov-file#single-header), change the code above to pass `true` to the `B3Propagator` constructor.

Author: Not required or supported on Cloud Foundry. I thought it was something that could previously be configured by Steeltoe and it's not obvious how it works now. But I looked again and I don't see it in source or docs, so we can probably just remove it.

## Exporting Distributed Traces

Previous versions of Steeltoe could automatically configure several different trace exporters, including [Zipkin](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Exporter.Zipkin), [OpenTelemetryProtocol (OTLP)](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Exporter.OpenTelemetryProtocol) and Jaeger.
The Jaeger exporter has been deprecated in favor of OTLP, which was only minimally configured by Steeltoe and is better described by [the official OTLP exporter documentation](https://opentelemetry.io/docs/languages/net/exporters/#otlp).

### Zipkin Server

To use the Zipkin Exporter, add a reference to the `OpenTelemetry.Exporter.Zipkin` NuGet package.

> Next, use the extension `AddZipkinExporter`:

Reviewer: ```suggestion
Next, use the extension method `AddZipkinExporter`:
```
